File: acp-c/standard_mix.c
/*
 * acp-c : Arcus C Client Performance benchmark program
 * Copyright 2013-2014 NAVER Corp.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <netinet/in.h>
#include <assert.h>

#include "libmemcached/memcached.h"
#include "common.h"
#include "keyset.h"
#include "valueset.h"
#include "client_profile.h"
#include "client.h"

static int
do_btree_test(struct client *cli)
{
  memcached_coll_create_attrs_st attr;
  memcached_return rc;
  int i, ok, keylen;
  uint64_t bkey;
  
  // Pick a key
  const char *key = keyset_get_key(cli->ks, NULL);
  keylen = strlen(key);
  
  // Create a btree item
  if (0 != client_before_request(cli))
    return -1;
  
  memcached_coll_create_attrs_init(&attr, 20 /* flags */, 100 /* exptime */,
    4000 /* maxcount */);
  memcached_coll_create_attrs_set_overflowaction(&attr,
    OVERFLOWACTION_SMALLEST_TRIM);
  rc = memcached_bop_create(cli->next_mc, key, keylen, &attr);
  ok = (rc == MEMCACHED_SUCCESS);
  if (!ok) {
    print_log("bop create failed. id=%d key=%s rc=%d(%s)", cli->id, key,
      rc, memcached_strerror(NULL, rc));
  }
  if (0 != client_after_request(cli, ok))
    return -1;
  
  // Insert a number of btree element
  for (i = 0; i < 100; i++) {
    uint8_t *val_ptr;
    int val_len;
    
    if (0 != client_before_request(cli))
      return -1;
    
    bkey = 1234 + i;

    val_ptr = NULL;
    val_len = valueset_get_value(cli->vs, &val_ptr);
    assert(val_ptr != NULL && val_len > 0 && val_len <= 4096);
    
    rc = memcached_bop_insert(cli->next_mc, key, keylen,
      bkey,
      NULL /* eflag */, 0 /* eflag length */,
      (const char*)val_ptr, (size_t)val_len,
      NULL /* Do not create automatically */);
    valueset_return_value(cli->vs, val_ptr);
    ok = (rc == MEMCACHED_SUCCESS);
    if (!ok) {
      print_log("bop insert failed. id=%d key=%s bkey=%llu rc=%d(%s)",
        cli->id, key, (long long unsigned)bkey,
        rc, memcached_strerror(NULL, rc));
    }
    if (0 != client_after_request(cli, ok))
      return -1;
  }
  
  // Update
  // Get/delete
  // Upsert
  // Incr/decr
  // Expire time
  // Smget
  // Getattr
  // PipedInsert
  return 0;
}

static int
do_set_test(struct client *cli)
{
  memcached_return rc;
  memcached_coll_create_attrs_st attr;
  int i, ok, keylen;
  const char *key;
  
  // Pick a key
  key = keyset_get_key(cli->ks, NULL);
  keylen = strlen(key);
  
  // Create a set item
  if (0 != client_before_request(cli))
    return -1;

  memcached_coll_create_attrs_init(&attr, 10 /* flags */, 100 /* exptime */,
    4000 /* maxcount */);
  memcached_coll_create_attrs_set_overflowaction(&attr, OVERFLOWACTION_ERROR);
  rc = memcached_sop_create(cli->next_mc, key, keylen, &attr);
  ok = (rc == MEMCACHED_SUCCESS);
  if (!ok) {
    print_log("sop create failed. id=%d key=%s rc=%d(%s)", cli->id, key,
      rc, memcached_strerror(NULL, rc));
  }
  if (0 != client_after_request(cli, ok))
    return -1;
  
  // Insert a number of elements.  Set has no element keys.
  for (i = 0; i < 100; i++) {
    uint8_t *val_ptr;
    int val_len;
    
    if (0 != client_before_request(cli))
      return -1;
    
    val_ptr = NULL;
    val_len = valueset_get_value(cli->vs, &val_ptr);
    assert(val_ptr != NULL && val_len > 0 && val_len <= 4096);

    rc = memcached_sop_insert(cli->next_mc, key, keylen,
      (const char*)val_ptr, (size_t)val_len,
      NULL /* Do not create automatically */);
    valueset_return_value(cli->vs, val_ptr);
    ok = (rc == MEMCACHED_SUCCESS);
    if (!ok) {
      print_log("sop insert failed. id=%d key=%s val_len=%d rc=%d(%s)",
        cli->id, key, val_len, rc, memcached_strerror(NULL, rc));
    }
    if (0 != client_after_request(cli, ok))
      return -1;
  }
  
  // Get/delete
  // Delete
  // PipedInsertBulk
  // Expire time
  // Getattr
  return 0;
}

static int
do_list_test(struct client *cli)
{
  memcached_return rc;
  memcached_coll_create_attrs_st attr;
  int i, ok, keylen;
  const char *key;
  
  // Pick a key
  key = keyset_get_key(cli->ks, NULL);
  keylen = strlen(key);
    
  // Create a list item
  if (0 != client_before_request(cli))
    return -1;
  
  memcached_coll_create_attrs_init(&attr, 10 /* flags */, 100 /* exptime */,
    4000 /* maxcount */);
  memcached_coll_create_attrs_set_overflowaction(&attr,
    OVERFLOWACTION_TAIL_TRIM);
  rc = memcached_lop_create(cli->next_mc, key, keylen, &attr);
  ok = (rc == MEMCACHED_SUCCESS);
  if (!ok) {
    print_log("lop create failed. id=%d key=%s rc=%d(%s)", cli->id, key,
      rc, memcached_strerror(NULL, rc));
  }
  if (0 != client_after_request(cli, ok))
    return -1;
  
  // Insert a number of elements.  Push at the head.
  for (i = 0; i < 100; i++) {
    uint8_t *val_ptr;
    int val_len;
    
    if (0 != client_before_request(cli))
      return -1;
    
    val_ptr = NULL;
    val_len = valueset_get_value(cli->vs, &val_ptr);
    assert(val_ptr != NULL && val_len > 0 && val_len <= 4096);

    rc = memcached_lop_insert(cli->next_mc, key, keylen,
      0 /* 0=head */,
      (const char*)val_ptr, (size_t)val_len,
      NULL /* Do not create automatically */);
    valueset_return_value(cli->vs, val_ptr);
    ok = (rc == MEMCACHED_SUCCESS);
    if (!ok) {
      print_log("lop insert failed. id=%d key=%s rc=%d(%s)", cli->id, key,
        rc, memcached_strerror(NULL, rc));
    }
    if (0 != client_after_request(cli, ok))
      return -1;
  }
  
  // Get/delete
  // Delete
  // PipedInsert
  // Expire time
  // Getattr
  return 0;
}

static int
do_simple_test(struct client *cli)
{
  memcached_return rc;
  int i, ok, keylen;
  const char *key;
  uint8_t *val_ptr;
  int val_len;
  
  // Set a number of items
  for (i = 0; i < 100; i++) {
    if (0 != client_before_request(cli))
      return -1;
    
    // Pick a key
    key = keyset_get_key(cli->ks, NULL);
    keylen = strlen(key);

    // Pick a value
    val_ptr = NULL;
    val_len = valueset_get_value(cli->vs, &val_ptr);
    assert(val_ptr != NULL && val_len > 0);
    
    rc = memcached_set(cli->next_mc, key, keylen, (const char*)val_ptr,
      (size_t)val_len, 100 /* exptime */, 0 /* flags */);
    valueset_return_value(cli->vs, val_ptr);
    ok = (rc == MEMCACHED_SUCCESS);
    if (!ok) {
      print_log("set failed. id=%d key=%s rc=%d(%s)", cli->id, key,
        rc, memcached_strerror(NULL, rc));
    }
    if (0 != client_after_request(cli, ok))
      return -1;
  }
  
  // Incr/decr
  // Get/delete
  // Update
  // Multiget
  // Expire time
  // Getattr
  // Cas
  // Add
  // Replace
  return 0;
}

static int
do_test(struct client *cli)
{
  if (0 != do_btree_test(cli))
    return -1;
  
  if (0 != do_set_test(cli))
    return -1;
  
  if (0 != do_list_test(cli))
    return -1;
  
  if (0 != do_simple_test(cli))
    return -1; // Stop the test
  
  return 0; // Do another test
}

static struct client_profile default_profile = {
  .do_test = do_test,
};

struct client_profile *
standard_mix_init(void)
{
  return &default_profile;
}

File: acp-java/standard_mix.java
/* -*- Mode: Java; tab-width: 2; c-basic-offset: 2; indent-tabs-mode: nil -*- */
/*
 * acp-java : Arcus Java Client Performance benchmark program
 * Copyright 2013-2014 NAVER Corp.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import java.util.concurrent.Future;
import java.util.concurrent.TimeUnit;

import net.spy.memcached.collection.CollectionAttributes;
import net.spy.memcached.collection.CollectionType;
import net.spy.memcached.collection.CollectionOverflowAction;
import net.spy.memcached.collection.ElementValueType;
import net.spy.memcached.internal.CollectionFuture;

// Should make subclasses of client instead of using the interface?
// FIXME

public class standard_mix implements client_profile {
  public boolean do_test(client cli) {
    try {
      if (!do_btree_test(cli))
        return false;
      
      if (!do_set_test(cli))
        return false;
    
      if (!do_list_test(cli))
        return false;
      
      if (!do_simple_test(cli))
        return false;
    } catch (Exception e) {
      System.out.printf("client_profile exception. id=%d exception=%s\n", 
                        cli.id, e.toString());
      e.printStackTrace();
    }
    return true;
  }
  
  private boolean do_delete_if_exist(client cli, String key, CollectionType type)
      throws Exception {
    // Get attributes
    if (!cli.before_request())
      return false;
    CollectionFuture<CollectionAttributes> f = cli.next_ac.asyncGetAttr(key);
    CollectionAttributes attr = f.get(1000L, TimeUnit.MILLISECONDS);
    if (!cli.after_request(true))
      return false;

    if (attr != null && (type != CollectionType.kv || type != attr.getType())) {
      if (!cli.before_request())
        return false;
      Future<Boolean> fb = cli.next_ac.delete(key);
      boolean ok = fb.get(1000L, TimeUnit.MILLISECONDS);
      if (!ok) {
        System.out.printf("KeyValue: delete failed. id=%d key=%s\n",
                          cli.id, key);
      }
      if (!cli.after_request(ok))
        return false;
    }
    return true; 
  }
 
  public boolean do_btree_test(client cli) throws Exception {
    // Pick a key
    String key = cli.ks.get_key();

    if (!do_delete_if_exist(cli, key, CollectionType.btree))
      return false;

    // Create a btree item
    if (!cli.before_request())
      return false;
    ElementValueType vtype = ElementValueType.BYTEARRAY;
    CollectionAttributes attr = 
      new CollectionAttributes(cli.conf.client_exptime,
                               CollectionAttributes.DEFAULT_MAXCOUNT,
                               CollectionOverflowAction.smallest_trim);
    CollectionFuture<Boolean> fb = cli.next_ac.asyncBopCreate(key, vtype, attr);
    boolean ok = fb.get(1000L, TimeUnit.MILLISECONDS);
    if (!ok) {
      System.out.printf("bop create failed. id=%d key=%s: %s\n", cli.id,
                        key, fb.getOperationStatus().getResponse());
    }
    if (!cli.after_request(ok))
      return false;

    // Insert a number of btree element
    cli.bks.reset();
    for (int i = 0; i < cli.bks.get_size(); i++) {
      if (!cli.before_request())
        return false;
      long bkey = cli.bks.get_bkey();
      byte[] val = cli.vset.get_value();
      assert(val.length <= 4096);
      fb = cli.next_ac.asyncBopInsert(key, bkey, null /* eflag */,
                                      val,
                                      null /* Do not auto-create item */);
      ok = fb.get(1000L, TimeUnit.MILLISECONDS);
      if (!ok) {
        System.out.printf("bop insert failed. id=%d key=%s bkey=%d: %s\n",
                          cli.id, key, bkey,
                          fb.getOperationStatus().getResponse());
      }
      if (!cli.after_request(ok))
        return false;
    }

    // Update
    // Get/delete
    // Upsert
    // Incr/decr
    // Expire time
    // Smget
    // Getattr
    // PipedInsert
    return true;
  }

  public boolean do_set_test(client cli) throws Exception {
    // Pick a key
    String key = cli.ks.get_key();
    
    if (!do_delete_if_exist(cli, key, CollectionType.set))
      return false;

    // Create a set item
    if (!cli.before_request())
      return false;
    ElementValueType vtype = ElementValueType.BYTEARRAY;
    CollectionAttributes attr = 
      new CollectionAttributes(cli.conf.client_exptime,
                               CollectionAttributes.DEFAULT_MAXCOUNT,
                               CollectionOverflowAction.error);
    // For set items, OverflowAction is always "error".

    CollectionFuture<Boolean> fb = cli.next_ac.asyncSopCreate(key, vtype, attr);
    boolean ok = fb.get(1000L, TimeUnit.MILLISECONDS);
    if (!ok) {
      System.out.printf("sop create failed. id=%d key=%s: %s\n", cli.id,
                        key, fb.getOperationStatus().getResponse());
    }
    if (!cli.after_request(ok))
      return false;
    
    // Insert a number of elements.  Set has no element keys.
    for (int i = 0; i < 100; i++) {
      if (!cli.before_request())
        return false;
      byte[] val = cli.vset.get_value();
      assert(val.length <= 4096);
      fb = cli.next_ac.asyncSopInsert(key, val,
                                      null /* Do not auto-create item */);
      ok = fb.get(1000L, TimeUnit.MILLISECONDS);
      if (!ok) {
        System.out.printf("sop insert failed. id=%d key=%s: %s\n", cli.id,
                          key, fb.getOperationStatus().getResponse());
      }
      if (!cli.after_request(ok))
        return false;
    }

    // Get/delete
    // Delete
    // PipedInsertBulk
    // Expire time
    // Getattr
    return true;
  }

  public boolean do_list_test(client cli) throws Exception {
    // Pick a key
    String key = cli.ks.get_key();
    
    if (!do_delete_if_exist(cli, key, CollectionType.list))
      return false;

    // Create a list item
    if (!cli.before_request())
      return false;
    ElementValueType vtype = ElementValueType.BYTEARRAY;
    CollectionAttributes attr = 
      new CollectionAttributes(cli.conf.client_exptime,
                               CollectionAttributes.DEFAULT_MAXCOUNT,
                               CollectionOverflowAction.tail_trim);
    // OverflowAction should be error, head_trim, or tail_trim
    CollectionFuture<Boolean> fb = cli.next_ac.asyncLopCreate(key, vtype, attr);
    boolean ok = fb.get(1000L, TimeUnit.MILLISECONDS);
    if (!ok) {
      System.out.printf("lop create failed. id=%d key=%s: %s\n", cli.id,
                        key, fb.getOperationStatus().getResponse());
    }
    if (!cli.after_request(ok))
      return false;

    // Insert a number of elements.  Push at the head.
    for (int i = 0; i < 100; i++) {
      if (!cli.before_request())
        return false;
      byte[] val = cli.vset.get_value();
      assert(val.length <= 4096);
      fb = cli.next_ac.asyncLopInsert(key, 0 /* head */, val,
                                  null /* Do not auto-create item */);
      ok = fb.get(1000L, TimeUnit.MILLISECONDS);
      if (!ok) {
        System.out.printf("lop insert failed. id=%d key=%s: %s\n", cli.id,
                          key, fb.getOperationStatus().getResponse());
      }
      if (!cli.after_request(ok))
        return false;
    }

    // Get/delete
    // Delete
    // PipedInsert
    // Expire time
    // Getattr
    return true;
  }

  public boolean do_simple_test(client cli) throws Exception {

    // Set a number of items
    for (int i = 0; i < 100; i++) {
      // Pick a key
      String key = cli.ks.get_key();
      if (!do_delete_if_exist(cli, key, CollectionType.btree))
        return false;

      if (!cli.before_request())
        return false;
      byte[] val = cli.vset.get_value();
      Future<Boolean> fb = 
        cli.next_ac.set(key, cli.conf.client_exptime, val, raw_transcoder.raw_tc);
      boolean ok = fb.get(1000L, TimeUnit.MILLISECONDS);
      if (!ok) {
        System.out.printf("set failed. id=%d key=%s\n", cli.id, key);
      }
      if (!cli.after_request(ok))
        return false;
    }

    // Incr/decr
    // Get/delete
    // Update
    // Multiget
    // Expire time
    // Getattr
    // Cas
    // Add
    // Replace
    return true;
  }
}
